pingcap/tidb
# util: add `affinity-cpus` starting argument

### What problem does this PR solve? 

when TiDB deployed in NUMA machine, we often need startup multiple processes and self-binding to numa node's cpu, so we should do this.

1. run `lscpu` to get info like this:

```
NUMA node0 CPU(s):     0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38
NUMA node1 CPU(s):     1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39
```

2. run `taskset` to bind core and start tidb-server

```
nohup taskset -c 0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38 bin/tidb-server
nohup taskset -c 1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39 bin/tidb-server
```

3. modify `max-procs` to meet bind count


the question's (2)(3) is not easy to use and often forgot to do (3)

### What is changed and how it works?

make (2)(3) easier, add an optional argument when TiDB-server start.

just run this:

```
./bin/tidb-server -store tikv -path 127.0.0.1:2379 -affinity-cpus=2,3
```

to bind proc to cpu 2 and 3, it will also be base step to make tidb-ansible to make deploy easier after ansible detect (1)

Notice: this argument only works on linux.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

run 
```
./bin/tidb-server -store tikv -path 127.0.0.1:2379 -affinity-cpus=2,3
```
take a look `taskset` to check bind works. 

```
taskset -pc $(pgrep tidb-server)
pid 14881's current affinity list: 2,3
```

run sysbench point-get and see `htop` only those cores be 100%

Code changes

 - N/A

Side effects

 - N/A

Related changes

 - Need to cherry-pick to 3.0


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/10773)
